src/dynarec/dynarec_arm_d9.c

```diff
@@ -281,39 +281,100 @@ uintptr_t dynarecD9(dynarec_arm_t* dyn, uintptr_t addr, uintptr_t ip, int ninst,
+                if(!box86_dynarec_fastround) {
+                    VCMP_F64_0(v1);
+                    B_MARK(cGE);           // if ST0 < 0 and if the rounding mode is toward 0, then use upward
+                    TSTS_IMM8(x2, 0b01);
+                    AND_IMM8_COND(cNE, x2, x2, 0b01); // 11 (TOWARDZERO) -> 01 (UPWARD), 01 -> 01
+                    MARK;
+                    BFI(x3, x2, 22, 2);     // inject new round
+                    VMSR(x3);
+                }
+
+                VMOV_64(0, v1);
+                CALL_1D(exp2, box86_dynarec_fastround ? 0 : (1 << x14));   // return is d0
                 VMOV_64(v1, 0);
                 addr+=4;
             } else {
-                VMOV_i_64(v1, 0b01110000);  // 1.0
-                VSUB_F64(v1, 0, v1);
+                if(!box86_dynarec_fastround) {
+                    BFI(x3, x2, 22, 2);     // inject new round
+                    VMSR(x3);               // put new fpscr
+                }
+
+                //ST0.d = expm1(LN2 * ST0.d);
+                MOV32(x2, (&d_ln2));
+                VLDR_64(0, x2, 0);
+                VMUL_F64(0, 0, v1);
+                x87_setround(dyn, ninst, x1, x2, -1);
+                CALL_1D(expm1, box86_dynarec_fastround ? 0 : (1 << x14));   // return is d0
+                VMOV_64(v1, 0);
             }
+            if(!box86_dynarec_fastround)
+                VMSR(x14);
             #endif
             // should set C1 to 0
             break;
         case 0xF1:
             INST_NAME("FYL2X");
             v1 = x87_get_st(dyn, ninst, x1, x2, 0, NEON_CACHE_ST_D);
             v2 = x87_get_st(dyn, ninst, x1, x2, 1, NEON_CACHE_ST_D);
+
+            if(!box86_dynarec_fastround) {
+                LDRH_IMM8(x1, xEmu, offsetof(x86emu_t, cw));    // hopefully cw is not too far for an imm8
+                UBFX(x1, x1, 10, 2);    // extract round...
+                VCMP_F64_0(v2);
+                VMRS_APSR();
+                B_MARK(cLT);            // if ST1.d < 0 then don't swap bits 0 and 1
+                BFI(x1, x1, 2, 1);      // if ST1.d >= 0 then swap bits 0 and 1
+                UBFX(x1, x1, 1, 2);
+                MARK;
+                s0 = fpu_get_scratch_double(dyn);
+                VMOV_i_64(s0, 0b01110000);   // = 1.0
+                VCMP_F64(v1, s0);
+                VMRS_APSR();
+                B_MARK2(cGE);           // if ST0 < 1 and if the rounding mode is toward 0, then use upward
+                TSTS_IMM8(x1, 0b01);
+                AND_IMM8_COND(cNE, x1, x1, 0b01); // 11 (TOWARDZERO) -> 01 (UPWARD), 01 -> 01
+                MARK2;
+                VMRS(x14);              // get fpscr
+                MOV_REG(x3, x14);
+                BFI(x3, x1, 22, 2);    // inject new round
+                VMSR(x3);               // put new fpscr
+            }
+
             VMOV_64(0, v1);    // prepare call to log2
-            CALL_1D(log2, 0);
+            CALL_1D(log2, box86_dynarec_fastround ? 0 : (1 << x14));
```

Comment:
this looks wrong. log2 doesn't return any integer, why trying to return stuff in x14?

Reply:
seems that the second argument of CALL_1D is the mask of registers which should be preserved?
https://github.com/ptitSeb/box86/blob/4861b78dc4f84359cf696824593e106035aa29e1/src/dynarec/dynarec_arm_helper.c#L453
https://github.com/ptitSeb/box86/blob/4861b78dc4f84359cf696824593e106035aa29e1/src/dynarec/dynarec_arm_helper.h#L258

Reply:
ah, ok, x14 has the old rounding value.